Invoke xcodebuild with CODE_SIGNING_* flags to avoid hanging

arthurtoper opened the pull request:
Hi,

We are using Slather in a fairly bespoke CI environment where unfortunately it's occasionally been hanging indefinitely when shelling out to xcodebuild. This PR implementation tweak using Open3.popen2e seems to fix the issue for us; we've not seen a single hang since making the change. However, I'm by no means a Ruby stream / I/O expert so if someone could suggest a better implementation I'll very happily adopt it.

I don't believe this change introduces any problems - certainly existing specs all still pass - but I'd appreciate it if others could give it a test too. I've not added any new specs to cover this as it's a pretty non-functional change, but happy to look into that if required.

Many thanks,

Arthur

EDIT - Forgot to mention, I'm pretty sure my version does not suppress STDERR as the original does (with the 'redirect stderr to avoid xcodebuild errors being printed' comment). Hope this isn't too much of a problem.

ksuther commented:
I haven't ever run into this hang myself, but you might try the workaround suggested in http://www.openradar.me/radar?id=5549442786656256 as an alternative.

neonichu commented on lib/slather/project.rb:
```diff
@@ -79,12 +80,19 @@ def derived_data_path
-      # redirect stderr to avoid xcodebuild errors being printed.
-      build_settings = `xcodebuild #{projectOrWorkspaceArgument} #{schemeArgument} -showBuildSettings #{buildAction} 2>&1`
+      build_settings = ''
+      Open3.popen2e('xcodebuild', projectOrWorkspaceArgument, schemeArgument, '-showbuildsettings', buildAction) do |stdin, stdout_err, wait_thr|
```
At least with Xcode 8 I'm getting an error when not specifying the argument in camel case, like `-showBuildSettings`

arthurtoper replied:
Good spot; that's my mistake, and I'll correct if ksuther's suggestion doesn't work.

neonichu commented on lib/slather/project.rb:
```diff
@@ -79,12 +80,19 @@ def derived_data_path
-        schemeArgument = nil
-        buildAction = nil
+        schemeArgument = ''
+        buildAction = ''
```
This change shouldn't be necessary.

arthurtoper commented:
ksuther: That workaround looks interesting and we're currently testing it out in our CI environment. Will report back once I have some news.

neonichu: Good spot; that's my mistake, and I'll correct if ksuther's suggestion doesn't work.

arthurtoper commented:
After fairly extensive testing, I'm happy to say that the workaround code signing flags suggested by ksuther appear to be working in our CI environment. We've not had a single hang since using those with the original backtick-style xcodebuild invocation. When I removed them and reverted back to the original command, the hangs reappeared immediately.

I've updated this PR with those flags.

bootstraponline commented:
This looks good to me.

ksuther commented:
Looks good. Sounds like this was the root cause of #234, #212, and #197 as well.

bootstraponline commented:
I think once the changelog is updated then this is ready to merge.

arthurtoper commented:
Thanks all. Do I need to do anything with the changelog, or would it be managed by a maintainer? Also happy to change the name of this PR to more accurately reflect its current contents.

neonichu commented:
Yah, please add an entry to the changelog for your change. Updating the PR title would also make sense.

arthurtoper commented:
I have updated the changelog and PR title. In the changelog I referenced the three issues which ksuther mentioned - hope that's okay, even though I wasn't specifically targeting them with the original PR. Let me know if I should change this! Thanks.

neonichu commented:
LGTM, thanks for your contribution 🎉
